lib/tapioca/compilers/dsl/active_support_concern.rb

```diff
@@ -0,0 +1,115 @@
+# typed: strict
+# frozen_string_literal: true
+
+require "tapioca/compilers/sorbet"
+
+begin
+  require "active_support"
+rescue LoadError
+  return
+end
+
+return unless Tapioca::Compilers::Sorbet.supports?(:mixes_in_class_methods_multiple_args)
+
+module Tapioca
+  module Compilers
+    module Dsl
+      # `Tapioca::Compilers::Dsl::ActiveSupportConcern` generates RBI files for classes that both `extend`
+      # `ActiveSupport::Concern` and `include` another class that extends `ActiveSupport::Concern`
+      #
+      # For example for the following hierarchy:
+      #
+      # ~~~rb
+      # # concern.rb
+      # module Foo
+      #  extend ActiveSupport::Concern
+      #  module ClassMethods; end
+      # end
+      #
+      # module Bar
+      #  extend ActiveSupport::Concern
+      #  module ClassMethods; end
+      #  include Foo
+      # end
+      #
+      # class Baz
+      #  include Bar
+      # end
+      # ~~~
+      #
+      # this generator will produce the RBI file `concern.rbi` with the following content:
+      #
+      # ~~~rbi
+      # # typed: true
+      # module Bar
+      #   mixes_in_class_methods(::Foo::ClassMethods)
+      # end
+      # ~~~
+      class ActiveSupportConcern < Base
+        extend T::Sig
+
+        sig { override.params(root: Parlour::RbiGenerator::Namespace, constant: Module).void }
+        def decorate(root, constant)
+          dependencies = linearized_dependencies_of(constant)
+
+          mixed_in_class_methods = dependencies
+            .uniq # Deduplicate
+            .map do |concern| # Map to class methods module name, if exists
+              "::#{name_of(concern)}::ClassMethods" if concern.const_defined?(:ClassMethods)
+            end
+            .compact # Remove non-existent records
```

Comment:
I generally find mutation of variables more problematic and harder to follow and understand, personally. There are places where it makes sense to avoid needless object creation but I would argue that the use case here is much less important in terms of overall performance. After all, this is a script that runs once and most of its time is spent loading the application in the first place.

I'd rather do a performance and optimization pass at a later date to find problematic code paths that do too much or allocate too much and prune them. Until then, I feel like it is a bit of a premature optimization in this use-case.